smartsim/settings/dispatch.py

```diff
@@ -0,0 +1,290 @@
+            )
+        return self.create_adapter_from_launcher(launcher, with_settings)
+
+
+@t.final
+class _LauncherAdapter(t.Generic[Unpack[_Ts]]):
+    def __init__(
+        self,
+        launcher: LauncherProtocol[_LaunchableT],
+        map_: t.Callable[[Unpack[_Ts]], _LaunchableT],
+    ) -> None:
+        # NOTE: We need to cast off the `_LaunchableT` -> `Any` in the
+        #       `__init__` method signature to hide the transform from users of
+        #       this class. If possible, this type should not be exposed to
+        #       users of this class!
+        self._adapt: t.Callable[[Unpack[_Ts]], t.Any] = map_
+        self._adapted_launcher: LauncherProtocol[t.Any] = launcher
+
+    def start(self, *args: Unpack[_Ts]) -> LaunchedJobID:
+        payload = self._adapt(*args)
+        return self._adapted_launcher.start(payload)
+
+
+DEFAULT_DISPATCHER: t.Final = Dispatcher()
+# Disabling because we want this to look and feel like a top level function,
+# but don't want to have a second copy of the nasty overloads
+# pylint: disable-next=invalid-name
+dispatch: t.Final = DEFAULT_DISPATCHER.dispatch
+
+
+# >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+# TODO: move these to a common module under `smartsim._core.launcher`
+# -----------------------------------------------------------------------------
+
+
+def create_job_id() -> LaunchedJobID:
+    return LaunchedJobID(str(uuid.uuid4()))
+
+
+class ExecutableProtocol(t.Protocol):
+    def as_program_arguments(self) -> t.Sequence[str]: ...
+
+
+class LauncherProtocol(t.Protocol[_T_contra]):
+    def start(self, launchable: _T_contra, /) -> LaunchedJobID: ...
+    @classmethod
+    def create(cls, exp: Experiment, /) -> Self: ...
+
+
+def make_shell_format_fn(
+    run_command: str | None,
+) -> _FormatterType[LaunchArguments, t.Sequence[str]]:
+    def impl(
+        args: LaunchArguments, exe: ExecutableProtocol, _env: _EnvironMappingType
+    ) -> t.Sequence[str]:
+        return (
+            (
+                run_command,
+                *(args.format_launch_args() or ()),
+                "--",
+                *exe.as_program_arguments(),
+            )
+            if run_command is not None
+            else exe.as_program_arguments()
+        )
+
+    return impl
+
+
+class ShellLauncher:
+    """Mock launcher for launching/tracking simple shell commands"""
+
+    def __init__(self) -> None:
+        self._launched: dict[LaunchedJobID, sp.Popen[bytes]] = {}
+
+    def start(self, command: t.Sequence[str]) -> LaunchedJobID:
+        id_ = create_job_id()
+        exe, *rest = command
+        # pylint: disable-next=consider-using-with
+        self._launched[id_] = sp.Popen((helpers.expand_exe_path(exe), *rest))
+        return id_
+
+    @classmethod
+    def create(cls, _: Experiment) -> Self:
+        return cls()
+
+
+# <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
```

Comment:
Coment or merge remnant?

Reply:
Comment! Its to guard off the similar `# >>>>..` line 246. I want that whole block moved to a new `smartsim.launchers` module or `smartsim._core.launchers` module or something to that nature.

There is probably a good argument to be made that it should be done in this PR, but like I mentioned I'm a bit worried about the size. If possible I'd like to put up a ticket and deal with it in the (very near) future!

Reply:
To be addressed in a follow on ticket